packages/matter.js/src/model/logic/record-validation/constraint.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * @license
+ * Copyright 2022-2023 Project CHIP Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Constraint } from "../../aspects/index.js";
+import { Metatype } from "../../definitions/index.js";
+import { ValueModel } from "../../models/index.js";
+import { type ValidatorBuilder } from "./ValidatorBuilder.js";
+
+export function addConstraint(builder: ValidatorBuilder, model: ValueModel, constraint: Constraint) {
+    builder.addTest(generateConstraintExpr(model, constraint), "CONSTRAINT_VIOLATION", model, "Value disallowed by constraint");
+
+    function generateConstraintExpr(model: ValueModel, ast: Constraint.Ast): string {
+        if (ast.parts) {
+            return `(${ast.parts.map(part => generateConstraintExpr(model, part)).join(" || ")})`;
+        }
+        
+        const metatype = model.metatype;
+        switch (metatype) {
+            case Metatype.array:
+                let entryTest;
+                if (ast.entry && model.children[0]) {
+                    entryTest = generateConstraintExpr(model.children[0], ast.entry);
+                }
+                return generateRangeTest("v.length", ast, "Array.isArray(v)", entryTest);
+
+            case Metatype.bytes:
+                return generateRangeTest("v.length", ast, "v instanceof Uint8Array");
+
+            case Metatype.string:
+                return generateRangeTest("v.length", ast, "typeof v == 'string'");
```

Comment:
also here please generate typesafe comparisms (including in other files)

Reply:
OK.  Note that this is a compiler, the code is ephemeral.  One might argue comparing a string to a string is a type safe comparison 😉